flink-connector-debezium/src/main/java/com/alibaba/ververica/cdc/debezium/DebeziumSourceFunction.java

```diff
@@ -519,4 +535,46 @@ public LinkedMap getPendingOffsetsToCommit() {
     public boolean getDebeziumStarted() {
         return debeziumStarted;
     }
+
+    private Class<?> determineDatabase() {
+        boolean isCompatibleWithLegacy =
+                FlinkDatabaseHistory.isCompatible(StateUtils.retrieveHistory(engineInstanceName));
+        if (LEGACY_IMPLEMENTATION_VALUE.equals(properties.get(LEGACY_IMPLEMENTATION_KEY))) {
+            // specifies the legacy implementation but the state may be incompatible
+            if (isCompatibleWithLegacy) {
+                return FlinkDatabaseHistory.class;
+            } else {
+                throw new IllegalStateException(
+                        "Use the legacy implementation of the database history but FlnkDatabaseHistory can't load the state.");
+            }
+        } else if (FlinkDatabaseSchemaHistory.isCompatible(
+                StateUtils.retrieveHistory(engineInstanceName))) {
+            // tries the non-legacy and checks the state is compatible
+            return FlinkDatabaseSchemaHistory.class;
+        } else if (isCompatibleWithLegacy) {
+            // starts from the state created before
+            return FlinkDatabaseHistory.class;
+        } else {
+            // impossible
+            throw new IllegalStateException("Can't determine which DatabaseHistory to use.");
+        }
+    }
+
+    // ---------------------------------------------------------------------------------------
+
+    /** Utils to get/put/remove the history. */
+    public static final class StateUtils {
+        public static void registerHistory(
+                String engineName, Collection<HistoryRecord> engineHistory) {
```

Comment:
we should `synchronize` visit the data to ensure the data won't change when checkpoint ?